src/checkpoints.rs

```diff
@@ -0,0 +1,193 @@
+//! Simple data structure to keep track of checkpointed state.  It stores a map
+//! of forks to a type and parent forks.
+//!
+//! `latest` forks is a set of all the forks with no children.
+//!
+//! A trunk is the latest fork that is a parent all the `latest` forks.  If
+//! consensus works correctly, then latest should be pruned such that only one
+//! trunk exists within N links.
+
+use hashbrown::{HashMap, HashSet};
+
+pub struct Checkpoints<T> {
+    /// Stores a map from fork to a T and a parent fork
+    pub checkpoints: HashMap<u64, (T, u64)>,
+
+    /// The latest forks that have been added
+    pub latest: HashSet<u64>,
+}
+
+impl<T> Checkpoints<T> {
+    pub fn load(&self, fork: u64) -> Option<&(T, u64)> {
+        self.checkpoints.get(&fork)
+    }
+    pub fn store(&mut self, fork: u64, data: T, parent: u64) {
+        if fork <= parent {
+            panic!(
+                "fork: {}, parent: {} error: out of order or trivial cycle",
+                fork, parent
+            );
+        }
+        self.latest.remove(&parent);
+        self.latest.insert(fork);
+        self.checkpoints.insert(fork, (data, parent));
+    }
+    /// Given a base fork, and a maximum number, collect all the
+    /// forks starting from the base fork backwards
+    pub fn collect(&self, num: usize, mut fork: u64) -> Vec<(u64, &T)> {
+        let mut rv = vec![];
+        while rv.len() < num {
+            if let Some((val, parent)) = self.load(fork) {
+                rv.push((fork, val));
+                fork = *parent;
+            } else {
+                break;
+            }
+        }
+        rv
+    }
+    /// given a maximum depth, find the highest parent shared by all
+    ///  forks
+    pub fn trunk(&self, num: usize) -> Option<u64> {
```

Comment:
why not a method that does
```
/// returns the trunks, and an inverse dag
fn invert(&self) -> (HashSet<u64>, HashMap<u64, HashSet<u64>>)
```

Reply:
and then?  probably what you've written is more along the lines of what you want for bank_state?